### ISSUE TITLE
[ASR] amp training face nan error

### ISSUE THREAD

**Issue by HsunGong** [espnet/espnet#4674]
**Describe the bug**

ASR training with `use_amp=true` face critical error, as in the second epoch the training fails.
As shown in the log, the grad_scale decrease and underflow suddenly, when I print the weight, I found the `weight=nan` starts from the convsubsampling.

**Basic environments:**
 - OS information: Ubuntu x86-64
 - python version: 3.7.4
 - espnet version: espnet 0.10.5
 - pytorch version: pytorch 1.8.0, cuda=10.1

**Task information:**
 - Task: [ASR]
 - Recipe: [librispeech]
 - ESPnet2

**To Reproduce**

1, 4 2080ti gpu running librispeech recipe
2, https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/conf/tuning/train_asr_conformer10_hop_length160.yaml
3, no speed perturb is used

Only change: 
```yaml
accum_grad: 1 # for faster reproduction
batch_bins: 3500000*4 # for 2080ti
```

**Error logs**

```
[speec258f0000F1] 2022-09-29 03:50:11,980 (trainer:706) INFO: 1epoch:train:10739-11505batch: iter_time=2.428e-04, forward_time=0.142, loss_ctc=122.599, loss_att=104.745, acc=0.496, loss=110.101, backward_time=0.138, optim0_lr0=6.946e-04, train_time=0.497, grad_norm=63.297, grad_scale=2.048e+03, optim_step_time=0.102
[speec258f0000F1] 2022-09-29 03:56:33,122 (trainer:706) INFO: 1epoch:train:11506-12272batch: iter_time=2.491e-04, forward_time=0.142, loss_ctc=120.951, loss_att=100.548, acc=0.514, loss=106.669, backward_time=0.137, optim0_lr0=7.425e-04, train_time=0.496, grad_norm=59.918, grad_scale=2.048e+03, optim_step_time=0.102
[speec258f0000F1] 2022-09-29 04:00:13,623 (trainer:646) WARNING: The grad norm is inf. Skipping updating the model.
[speec258f0000F1] 2022-09-29 04:02:55,551 (trainer:706) INFO: 1epoch:train:12273-13039batch: iter_time=2.422e-04, forward_time=0.142, loss_ctc=124.157, loss_att=101.177, acc=0.529, loss=108.071, backward_time=0.139, optim0_lr0=7.904e-04, train_time=0.498, grad_norm=inf, grad_scale=2.659e+03, optim_step_time=0.102
[speec258f0000F1] 2022-09-29 04:09:16,186 (trainer:706) INFO: 1epoch:train:13040-13806batch: iter_time=2.457e-04, forward_time=0.142, loss_ctc=118.965, loss_att=94.877, acc=0.543, loss=102.103, backward_time=0.138, optim0_lr0=8.383e-04, train_time=0.495, grad_norm=53.294, grad_scale=2.048e+03, optim_step_time=0.101
[speec258f0000F1] 2022-09-29 04:15:36,632 (trainer:706) INFO: 1epoch:train:13807-14573batch: iter_time=2.413e-04, forward_time=0.142, loss_ctc=112.960, loss_att=88.854, acc=0.556, loss=96.086, backward_time=0.138, optim0_lr0=8.862e-04, train_time=0.495, grad_norm=49.579, grad_scale=2.048e+03, optim_step_time=0.101
[speec258f0000F1] 2022-09-29 04:17:34,075 (trainer:646) WARNING: The grad norm is inf. Skipping updating the model.
[speec258f0000F1] 2022-09-29 04:22:00,212 (trainer:706) INFO: 1epoch:train:14574-15340batch: iter_time=2.388e-04, forward_time=0.142, loss_ctc=113.460, loss_att=87.773, acc=0.569, loss=95.479, backward_time=0.139, optim0_lr0=9.341e-04, train_time=0.499, grad_norm=inf, grad_scale=2.294e+03, optim_step_time=0.103
[speec258f0000F1] 2022-09-29 04:23:54,562 (trainer:329) INFO: 1epoch results: [train] iter_time=2.672e-04, forward_time=0.142, loss_ctc=169.600, loss_att=154.618, acc=0.333, loss=159.112, backward_time=0.139, optim0_lr0=4.794e-04, train_time=0.498, grad_norm=inf, grad_scale=1.586e+03, optim_step_time=0.102, time=2 hours, 7 minutes and 40.36 seconds, total_count=15357, gpu_max_cached_mem_GB=14.377, [valid] loss_ctc=91.082, cer_ctc=0.378, loss_att=72.874, acc=0.500, cer=0.391, wer=0.992, loss=78.336, time=48.88 seconds, total_count=185, gpu_max_cached_mem_GB=14.377, [att_plot] time=56.09 seconds, total_count=0, gpu_max_cached_mem_GB=14.377
[speec258f0000F1] 2022-09-29 04:23:59,616 (trainer:377) INFO: The best model has been updated: valid.acc
[speec258f0000F1] 2022-09-29 04:23:59,616 (trainer:263) INFO: 2/50epoch started. Estimated time to finish: 4 days, 9 hours and 45 minutes
[speec258f0000F1] 2022-09-29 04:30:24,597 (trainer:706) INFO: 2epoch:train:1-767batch: iter_time=6.354e-04, forward_time=0.141, loss_ctc=109.188, loss_att=82.599, acc=0.582, loss=90.576, backward_time=0.138, optim_step_time=0.100, optim0_lr0=9.831e-04, train_time=0.501, grad_norm=45.189, grad_scale=1.024e+03
[speec258f0000F1] 2022-09-29 04:36:44,664 (trainer:706) INFO: 2epoch:train:768-1534batch: iter_time=2.374e-04, forward_time=0.141, loss_ctc=108.499, loss_att=81.789, acc=0.586, loss=89.802, backward_time=0.139, optim_step_time=0.101, optim0_lr0=0.001, train_time=0.495, grad_norm=43.223, grad_scale=1.024e+03
[speec258f0000F1] 2022-09-29 04:43:06,991 (trainer:706) INFO: 2epoch:train:1535-2301batch: iter_time=2.323e-04, forward_time=0.141, loss_ctc=110.885, loss_att=82.577, acc=0.596, loss=91.069, backward_time=0.139, optim_step_time=0.103, optim0_lr0=0.001, train_time=0.498, grad_norm=42.126, grad_scale=1.447e+03
[speec258f0000F1] 2022-09-29 04:44:25,398 (trainer:646) WARNING: The grad norm is inf. Skipping updating the model.
[speec258f0000F1] 2022-09-29 04:48:57,252 (trainer:646) WARNING: The grad norm is nan. Skipping updating the model.
[speec258f0000F1] 2022-09-29 04:49:08,298 (trainer:646) WARNING: The grad norm is nan. Skipping updating the model.
[speec258f0000F1] 2022-09-29 04:49:25,995 (trainer:706) INFO: 2epoch:train:2302-3068batch: iter_time=2.381e-04, forward_time=0.141, loss_ctc=108.827, loss_att=80.240, acc=0.602, loss=88.816, backward_time=0.137, optim_step_time=0.102, optim0_lr0=0.001, train_time=0.493, grad_norm=inf, grad_scale=1.184e+03
[speec258f0000F1] 2022-09-29 04:51:11,345 (trainer:646) WARNING: The grad norm is nan. Skipping updating the model.
...
[speec258f0000F1] 2022-09-29 04:55:44,915 (trainer:706) INFO: 2epoch:train:3069-3835batch: iter_time=2.356e-04, forward_time=0.141, loss_ctc=108.541, loss_att=79.342, acc=0.603, loss=88.102, backward_time=0.137, optim_step_time=0.101, optim0_lr0=0.001, train_time=0.493, grad_norm=40.503, grad_scale=108.360
[speec258f0000F1] 2022-09-29 04:56:02,917 (trainer:646) WARNING: The grad norm is nan. Skipping updating the model.
...
[speec258f0000F1] 2022-09-29 05:01:59,534 (trainer:706) INFO: 2epoch:train:3836-4602batch: iter_time=2.360e-04, forward_time=0.141, loss_ctc=186.858, loss_att=279.597, acc=0.363, loss=251.775, backward_time=0.138, optim_step_time=0.103, optim0_lr0=0.001, train_time=0.488, grad_norm=20.707, grad_scale=0.369
[speec258f0000F1] 2022-09-29 05:02:11,368 (trainer:646) WARNING: The grad norm is nan. Skipping updating the model.
...
[speec258f0000F1] 2022-09-29 05:07:57,900 (trainer:706) INFO: 2epoch:train:4603-5369batch: iter_time=2.368e-04, forward_time=0.141, loss_ctc=255.303, loss_att=490.458, acc=0.133, loss=419.911, backward_time=0.139, optim_step_time=0.101, optim0_lr0=0.001, train_time=0.466, grad_norm=0.000e+00, grad_scale=1.928e-21
```

**Comment by sw005320** [comment]
Thanks for your report.
This level of instability might not be easily avoided. I'd recommend you change the optimization parameters (e.g., learning rate, batch size, accum grad) to find some lucky setup.
If you still have some issues and they are reproducible, please report them.

**Comment by HsunGong** [comment]
I was wondering if it is the `accum_grad` issue, and find that if `accum_grad=8` or more, the NaN error happened after 15 epochs.
---
The equal batch-bins is 

Original: 35000000 for 4 gpu with accum=4 (3.5e7)
NaN at 2epo: 3275000 for 1 gpu with accum=1 (3.2e6)
NaN at 14epo:  3275000*10 for 1 gpu with accum=10 (3.3e7)

So I wonder if the amp training NaN error has correlation with batch size, and if the batch size is too small, amp can not be used.

**Comment by sw005320** [comment]
Thanks for your report. This is very useful.
I also added @pyf98 to this discussion. @pyf98, can you give us a comment about it?

My concern about this is batch norm. As far as I remember, our accum grad does not consider the batch norm.
Even if we try to virtually increase the batch size by controlling accum grad, this batch norm part was conducted with the original batch size, which would make the estimation unstable (especially for second order statistics).
Again, maybe I’m wrong and want to ask @pyf98’s comments.

**Comment by pyf98** [comment]
@HsunGong Thanks for your question. I think it is caused by the inappropriate learning rate (and its schedule). In theory, the learning rate needs to be scaled along with the total batch size.

Our original total batch_bins is 35000000 * 4 which is much larger than your current value. Using the same learning rate with a much smaller batch size can cause divergence, which is even worse due to mixed precision.

So my suggestion is to tune the learning rate and warmup steps based on your own environment. Or try to always use a similar total batch_bins as our config. (However, as @sw005320 mentioned, using accum grad is not strictly equivalent.)

**Comment by pyf98** [comment]
Another suggestion (not related to amp issues though): I notice that you don't use speed perturbation. So the total number of epochs needs to be increased (3x) if you want to reproduce a similar result.

**Comment by HsunGong** [comment]
OK, I'll try it
----
I found that with speed perturb the final result can be boosted from 5.4 (reproduction) to 5.2 (offical), so maybe this is not necessary as we need to store more speech data.
